src/linalg.jl

```diff
@@ -215,20 +215,17 @@ end
 end
 
 @inline dot(a::StaticVector, b::StaticVector) = _vecdot(same_size(a, b), a, b)
-@generated function _vecdot(::Size{S}, a::StaticArray, b::StaticArray) where {S}
+@inline function _vecdot(::Size{S}, a::StaticArray, b::StaticArray) where {S}
     if prod(S) == 0
-        return :(zero(promote_op(*, eltype(a), eltype(b))))
-    end
-
-    expr = :(adjoint(a[1]) * b[1])
-    for j = 2:prod(S)
-        expr = :($expr + adjoint(a[$j]) * b[$j])
+        return zero(promote_op(*, eltype(a), eltype(b)))
     end
-
-    return quote
-        @_inline_meta
-        @inbounds return $expr
+    za = zero(a[1])
```

Comment:
Have you checked if this still works for block vectors, `SVector{Vector{<:Number}}`? (Part of the advantage of unrolling in this case was that you don't need to start with a "zero" element).

Reply:
Yes, that case is already being tested for:

https://github.com/JuliaArrays/StaticArrays.jl/blob/d3f739ebd3a465ee014ca53ac893616704318a3a/test/linalg.jl#L120

Reply:
Reverting to the `2 : prod(S)` loop results in the same performance in benchmark 2 up to `n = 30` (the point at which the loop is no longer unrolled). After that point, there's a significant performance drop compared to the PR branch (up to 2x for `n = 32`!).